Networking parameters manager upgrade.

shamil-gadelshin opened the pull request:
This PR improves networking parameters manager for the `subspace-networking` crate.
Original issue: https://github.com/subspace/subspace/issues/722

#### Changes
- simplified examples  (keypair pregeneration)
- add `BootstrappedNetworkingParameters` stub and removed NetworkingParametersRegistryStub`
- remove bootstrap addresses from Kademlia initialization
- add `remove_known_peer_addresses` method and use it on `OutgoingConnnectionError` swarm event
- add dialing to known peers
- add `established connections` swarm limit
- added time-based expiration for non-responding peers

#### Open questions
- `swarm.dial()` is asynchronous despite its API and doesn't return connection error immediately, I limit all established connections instead of trying addresses for connectivity batch after batch. However, I'm not sure it's a good idea and keep it for now as an example. We could limit initial connections by limiting initial cache size or introduce an additional async-worker to dial known addresses in batches. 

@nazar-pc  Please, comment here.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

shamil-gadelshin committed:
networking: Simplify examples - add explicit keypair.

shamil-gadelshin committed:
Introduce BootstrappedNetworkingParameters for networking.

shamil-gadelshin committed:
networking: Add explicit boostrap addresses dialing.

shamil-gadelshin committed:
networking: Add `remove_known_peer_addresses` method.

- introduce connection limit

shamil-gadelshin committed:
networking: Add last failure time to address cache.

nazar-pc commented on crates/subspace-networking/examples/get-peers-complex.rs:
```diff
@@ -1,3 +1,4 @@
+use crate::identity::sr25519::Keypair;
```
`crate::identity` looks like re-export, better use `libp2p::identity` if that is the case, helps a lot to keep imports clean in each individual module.

nazar-pc commented on crates/subspace-networking/examples/bootstrap-node.rs:
```diff
@@ -15,7 +15,7 @@ enum Command {
         /// Multiaddrs of bootstrap nodes to connect to on startup, multiple are supported
         #[clap(long)]
-        bootstrap_node: Vec<Multiaddr>,
+        bootstrap_nodes: Vec<Multiaddr>,
```
It was named that way such that you can use `--bootstrap-node /x/y/z` in CLI. I guess we can rename it here with an attribute to keep both CLI argument and property name convenient.

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -58,6 +62,48 @@ impl NetworkingParametersRegistry for NetworkingParametersRegistryStub {
+#[derive(Clone)]
+pub struct BootstrappedNetworkingParameters {
+    bootstrap_addresses: Vec<Multiaddr>,
+}
```
Why do we need this? It seems to be just like stub, except stub doesn't accept bootstrap nodes as an argument, but it could.

shamil-gadelshin replied:
We removed `bootstrap_addresses` from Kademlia, it means that we need a stub with basic bootstrap addresses provider feature for examples.

nazar-pc replied:
Why not combine those two though? Stub is the same as this one with an empty list of addresses, right? Also if it is just for examples, it should probably live there.

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -287,3 +344,30 @@ impl NetworkingParameters {
+            if let Some(peer_id) = peer_id {
+                Some((peer_id, modified_multiaddr))
+            } else {
+                debug!(%multiaddr, "Incorrect multiaddr provided for bootstrap");
```
This is not a debug message. If you don't want to return an error from here outright, this at least deserves a warning. Would be very confusing to debug if things silently don't work as expected.

nazar-pc commented on crates/subspace-networking/src/request_handlers/generic_request_handler.rs:
```diff
@@ -60,6 +60,12 @@ impl<Request: GenericRequest> GenericRequestHandler<Request> {
+    /// registration in the libp2p stack in client-only usage scenarios.
+    pub fn client_only() -> Box<dyn RequestHandler> {
+        Self::create(|_| None)
+    }
```
Abstraction naming is completely broken here. You create a generic request handler in a mode that doesn't handle any requests.

Things need to be renamed in some way that makes this nicer. I understand why you did this, but I'd use `::create()` for now until we figure out a nicer way to structure this.

Something like `Box<dyn RequestResponseHandler>` would make more sense in that case and other things need to be renamed accordingly, but better in a separate PR.

shamil-gadelshin replied:
I'm not sure why it's broken since we allow empty handlers to be provided and this is by design. However, I'll remove it for now if you consider it contraversial.

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -30,8 +30,7 @@ pub trait NetworkingParametersRegistry: Send {
     /// Returns known addresses from networking parameters DB. It removes p2p-protocol suffix.
-    /// Peer number parameter limits peers to retrieve.
-    async fn known_addresses(&self, peer_number: usize) -> Vec<(PeerId, Multiaddr)>;
+    async fn known_addresses(&self) -> Vec<(PeerId, Multiaddr)>;
```
Hm... this can be quite a lot... I'm not sure this is a good API

shamil-gadelshin replied:
I changed it after our conversation to batch-based method.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -30,12 +31,12 @@ use std::time::Duration;
 const GOSSIPSUB_PROTOCOL_PREFIX: &str = "subspace/gossipsub";
 // Max bootstrap addresses to preload.
-const INITIAL_BOOTSTRAP_ADDRESS_NUMBER: usize = 100;
+const INITIAL_BOOTSTRAP_ADDRESS_LIMIT: usize = 200;
```
That feels excessive, why did you double the number?

shamil-gadelshin replied:
It is an intermediate value, not it's back to 100.

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -232,6 +232,27 @@ pub async fn create(config: Config) -> Result<(Node, NodeRunner), CreationError>
+                debug!(%err, %peer_id, %addr, "Bootstrap address dialed successfully." )
+            } else {
+                dialed_addresses_number += 1;
+                trace!(%peer_id, %addr, "Failed to dial a bootstrap address." )
```
If/else messages are wrong. You have an error field in success message and vice versa :confused: 

Also note that no error here doesn't mean at all that peer is successfully reached, it just means that the address is correct, peer is not banned and a few similar checks.

shamil-gadelshin replied:
I changed it in the last commit, however, the whole snippet is not relevant anymore.

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -251,6 +232,28 @@ impl NetworkingParametersRegistry for NetworkingParametersManager {
+                }
+            })
+            .for_each(|addr| {
+                if let Some(addresses) = self.known_peers.get_mut(&peer_id) {
```
`.get_mut` will update recency of the record in LRU, are you sure you didn't want to use `.peek_mut` instead?

shamil-gadelshin replied:
I agree.

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -235,19 +231,28 @@ impl NetworkingParametersRegistry for NetworkingParametersManager {
+                            // if we failed less than a day ago
+                            if time.add(chrono::Duration::days(1)) > Utc::now() {
+                                // Update failure time
+                                *last_failed_time = Some(Utc::now())
```
I feel like you don't need the whole `crono` crate for this, use `std::time::Instant` in idiomatic way:
```suggestion
                            // if we failed less than a day ago
                            if time.elapsed() < Duration::from_secs(3600 * 24) {
                                // Update failure time
                                last_failed_time.replace(Instant::now());
```

It will also prevent you from getting negative time due to `SystemTime` used by `Utc::now()` being non-monotonic (not sure if underflow will cause crash or ridiculously large time and I don't care), while `Instant` is monotonic and specifically designed for such purposes where you want to measure difference in time.

And lastly I don't think I fully understand why would you want to store last and not first failure time? If you retry over and over again last value will update and not give you any helpful information about what chances are that address is not going to be reachable ever again.

shamil-gadelshin replied:
We need `chrono` for its `serde` feature which is missing in the `std`. I'm not sure how `Utc::now() being non-monotonic` is affecting us here on the scale of days, please explain. 

The `the last` instead of `the first` is an error (and I even tested it locally !!!) , I refactored the method to a testable alternative and introduced a unit-test.

nazar-pc replied:
Do you need to serialize timestamps to disk at all BTW? It just feels like the wrong primitive for the job here.

> The the last instead of the first is an error (and I even tested it locally !!!) , I refactored the method to a testable alternative and introduced a unit-test.

I believe you that there could be some kind of error if you change it to the opposite, but it doesn't mean the business logic is correct. Can you describe with words what you wanted to achieve here?

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -312,32 +317,31 @@ impl NetworkingParametersRegistry for NetworkingParametersManager {
 
-            for address in address_set.iter().cloned() {
-                address_cache.push(address, ());
+            for (address, last_failed) in address_map.iter() {
+                address_cache.push(address.clone(), *last_failed);
```
One more nit while I'm at it:
```suggestion
            for (address, last_failed) in address_map.iter().cloned() {
                address_cache.push(address, last_failed);
```

shamil-gadelshin replied:
The suggestion is not compilable.

nazar-pc reviewed (commented):
I see what you're trying to achieve, but the only thing that has an effect is the fact that you're trying 200 peers. There is no guarantee that any of them are alive. And if none are, no recovery mechanism is present to ever get out of that situation.

What I expect is that we need to keep track of how bad situation is from number of active connections and react to that. Then you don't need to start with 200 peers (or you can, though number should eventually be configurable like in Substrate) and can query more and more peers from persistence layer (eventually falling back to bootstrap nodes) until node gets back online.

Also re-checking of peers that were previously considered offline is missing. There needs to be a mechanism for distinguishing whether it is node that is offline or individual peers.

shamil-gadelshin committed:
networking: Refactor connection dialing.

- introduce asynchronous batch dialing
- add CheckConnectedPeers Node command and wait_for_connected_peers() method
- fix review commnts

shamil-gadelshin committed:
networking: Fix remove_known_peer_addresses

shamil-gadelshin committed:
Merge branch 'main' into networking-parameters-manager-upgrade

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -9,24 +10,30 @@ use parity_db::{Db, Options};
 type LastFailedTime = Option<DateTime<Utc>>;
 
+// Convenience alias for peer ID and its multiaddresses.
+type PeerAddresses = (PeerId, Multiaddr);
```
This is just one address though

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -163,9 +170,29 @@ impl NetworkingParametersManager {
         })
     }
 
+    // Returns known addresses from networking parameters DB. It removes p2p-protocol suffix.
```
It doesn't remove anything and I'm not sure that is even helpful in the API docs, it is kind of obvious from the signature.

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -523,6 +557,15 @@ impl NodeRunner {
                 );
             }
+            Command::CheckConnectedPeers { result_sender } => {
+                let connected_peers_present = self.swarm.connected_peers().next().is_some();
```
FYI, there is `connected_peers_count` already in `Shared`, might be faster to access comparing to accessing concrete peers that you don't care about anyway.

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -112,6 +119,53 @@ impl NodeRunner {
                 _ = self.networking_parameters_registry.run().fuse() => {
                     trace!("Network parameters registry runner exited.")
+                },
+                _ = &mut self.peer_dialing_timeout => {
```
I think this should be combined with `random_query_timeout`, which tries to find peers on the network. Overall the idea I had was to monitor number of connections and respond to disconnections rather than having an interval running every 3 seconds.

shamil-gadelshin replied:
Added TODO as agreed

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -112,6 +119,53 @@ impl NodeRunner {
+        if connected_peers.len() < CONNECTED_PEERS_THRESHOLD {
+            debug!(
+                %local_peer_id,
+                "Initiate connection to known peers [connected peers={}]", connected_peers.len()
```
You could have use structured logs instead of trying to format it manually, especially since you have `local_peer_id` above already.

nazar-pc commented on crates/subspace-networking/src/node_runner.rs:
```diff
@@ -173,14 +227,7 @@ impl NodeRunner {
-                    }
-                };
                 debug!("Connection established with peer {peer_id} [{num_established} from peer]");
-                shared.connected_peers_count.fetch_add(1, Ordering::SeqCst);
```
Why did you remove it? It was one atomic read away in `Node`, now you need to make a request over the channel to know how many connected peers there is. We actually want more of these variables, not less IMO.

nazar-pc commented on crates/subspace-networking/src/utils.rs:
```diff
@@ -1,5 +1,11 @@
 use libp2p::multiaddr::Protocol;
 use libp2p::Multiaddr;
+use std::marker::PhantomData;
+
```
Extra new line

nazar-pc commented on crates/subspace-networking/src/behavior.rs:
```diff
@@ -1,5 +1,7 @@
 pub(crate) mod custom_record_store;
 pub(crate) mod persistent_parameters;
+#[cfg(test)]
+mod tests;
```
Modules above imports, all imports together, please

shamil-gadelshin replied:
The new module seems to be above imports 👀 .

nazar-pc replied:
My bad :facepalm:

nazar-pc commented on crates/subspace-networking/src/behavior/persistent_parameters.rs:
```diff
@@ -414,3 +409,48 @@ fn remove_p2p_suffix(address: Multiaddr) -> Multiaddr {
 }
+
+// Testable implementation of the `remove_known_peer_addresses`
+pub(super) async fn remove_known_peer_addresses_internal(
```
Doesn't need to be async.

nazar-pc reviewed (commented):
Please try to put distinct changes into separate commits, helps with review.

shamil-gadelshin committed:
networking: Fix review comments.

shamil-gadelshin committed:
Merge branch 'main' into networking-parameters-manager-upgrade
